In [50]:
print("Hola")

Hola


## VANGUARD 

In [51]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns 

In [52]:
df_exp_clients = pd.read_csv("df_final_experiment_clients.txt")
df_web1 = pd.read_csv("df_final_web_data_pt_1.txt")
df_web2 = pd.read_csv("df_final_web_data_pt_2.txt")
df_demo = pd.read_csv("df_final_demo.txt")

TAMAÑO DE CADA DATASET

In [53]:
#Filas y columnas de c/u

dfs = {
    "exp_clients": df_exp_clients,
    "web1": df_web1,
    "web2": df_web2,
    "demo": df_demo
}

for name, df in dfs.items():
    print(f"\n{name}")
    print("Shape:", df.shape)
    print("Columnas:", df.columns.tolist())



exp_clients
Shape: (70609, 2)
Columnas: ['client_id', 'Variation']

web1
Shape: (343141, 5)
Columnas: ['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']

web2
Shape: (412264, 5)
Columnas: ['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']

demo
Shape: (70609, 9)
Columnas: ['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr', 'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth']


  DATOS 

In [54]:

df_exp_clients.head()

client_id Variation
0    9988021      Test
1    8320017      Test
2    4033851   Control
3    1982004      Test
4    9294070   Control

In [55]:
df_web1.head()

client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04

In [56]:
df_web2.head()

client_id             visitor_id                      visit_id  \
0     763412  601952081_10457207388  397475557_40440946728_419634   
1    6019349  442094451_91531546617  154620534_35331068705_522317   
2    6019349  442094451_91531546617  154620534_35331068705_522317   
3    6019349  442094451_91531546617  154620534_35331068705_522317   
4    6019349  442094451_91531546617  154620534_35331068705_522317   

  process_step            date_time  
0      confirm  2017-06-06 08:56:00  
1      confirm  2017-06-01 11:59:27  
2       step_3  2017-06-01 11:58:48  
3       step_2  2017-06-01 11:58:08  
4       step_1  2017-06-01 11:57:58

In [57]:
df_demo.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1    2304905             7.0              94.0      58.0     U        2.0   
2    1439522             5.0              64.0      32.0     U        2.0   
3    1562045            16.0             198.0      49.0     M        2.0   
4    5126305            12.0             145.0      33.0     F        2.0   

         bal  calls_6_mnth  logons_6_mnth  
0   45105.30           6.0            9.0  
1  110860.30           6.0            9.0  
2   52467.79           6.0            9.0  
3   67454.65           3.0            6.0  
4  103671.75           0.0            3.0

In [58]:
#TODOS TIENEN EN COMÚN CLIENT_ID , ASÍ COMO VARIABLES CATEGORICAS Y NUMÉRICAS 

In [59]:
df_exp_clients.info()
df_web1.info()
df_web2.info()
df_demo.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   client_id  70609 non-null  int64 
 1   Variation  50500 non-null  object
dtypes: int64(1), object(1)
memory usage: 1.1+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 343141 entries, 0 to 343140
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   client_id     343141 non-null  int64 
 1   visitor_id    343141 non-null  object
 2   visit_id      343141 non-null  object
 3   process_step  343141 non-null  object
 4   date_time     343141 non-null  object
dtypes: int64(1), object(4)
memory usage: 13.1+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 412264 entries, 0 to 412263
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   client_id     412264 no

Tipos de Variables y Valores Nulos

In [60]:
for name, df in dfs.items():
    print(f"\n {name}")
    display(df.info())


 exp_clients
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   client_id  70609 non-null  int64 
 1   Variation  50500 non-null  object
dtypes: int64(1), object(1)
memory usage: 1.1+ MB


None


 web1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 343141 entries, 0 to 343140
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   client_id     343141 non-null  int64 
 1   visitor_id    343141 non-null  object
 2   visit_id      343141 non-null  object
 3   process_step  343141 non-null  object
 4   date_time     343141 non-null  object
dtypes: int64(1), object(4)
memory usage: 13.1+ MB


None


 web2
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 412264 entries, 0 to 412263
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   client_id     412264 non-null  int64 
 1   visitor_id    412264 non-null  object
 2   visit_id      412264 non-null  object
 3   process_step  412264 non-null  object
 4   date_time     412264 non-null  object
dtypes: int64(1), object(4)
memory usage: 15.7+ MB


None


 demo
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   client_id         70609 non-null  int64  
 1   clnt_tenure_yr    70595 non-null  float64
 2   clnt_tenure_mnth  70595 non-null  float64
 3   clnt_age          70594 non-null  float64
 4   gendr             70595 non-null  object 
 5   num_accts         70595 non-null  float64
 6   bal               70595 non-null  float64
 7   calls_6_mnth      70595 non-null  float64
 8   logons_6_mnth     70595 non-null  float64
dtypes: float64(7), int64(1), object(1)
memory usage: 4.8+ MB


None

In [61]:
#object deberian ser numércias...


In [62]:
#Variables categóricas, buscando IDs encubiertos, variables con pocas categorias y/o columnas basura 

In [63]:
df_exp_clients.select_dtypes(include="object").nunique()


Variation    2
dtype: int64

In [64]:
df_web1.select_dtypes(include="object").nunique()

visitor_id       62936
visit_id         75256
process_step         5
date_time       282666
dtype: int64

In [65]:
df_web2.select_dtypes(include="object").nunique()

visitor_id       71042
visit_id         82841
process_step         5
date_time       346697
dtype: int64

In [66]:
df_demo.select_dtypes(include="object").nunique()

gendr    4
dtype: int64

Estadística, buscando rangos fuera de lugar

In [67]:
for name, df in dfs.items():
    print(f"\n {name}")
    display(df.describe())



 exp_clients


client_id
count  7.060900e+04
mean   5.004992e+06
std    2.877278e+06
min    1.690000e+02
25%    2.519329e+06
50%    5.016978e+06
75%    7.483085e+06
max    9.999839e+06


 web1


client_id
count  3.431410e+05
mean   4.996097e+06
std    2.875839e+06
min    1.690000e+02
25%    2.506067e+06
50%    5.007915e+06
75%    7.468519e+06
max    9.999839e+06


 web2


client_id
count  4.122640e+05
mean   5.028227e+06
std    2.881828e+06
min    3.360000e+02
25%    2.533952e+06
50%    5.056314e+06
75%    7.516792e+06
max    9.999875e+06


 demo


client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  7.060900e+04    70595.000000      70595.000000  70594.000000   
mean   5.004992e+06       12.052950        150.659367     46.442240   
std    2.877278e+06        6.871819         82.089854     15.591273   
min    1.690000e+02        2.000000         33.000000     13.500000   
25%    2.519329e+06        6.000000         82.000000     32.500000   
50%    5.016978e+06       11.000000        136.000000     47.000000   
75%    7.483085e+06       16.000000        192.000000     59.000000   
max    9.999839e+06       62.000000        749.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  70595.000000  7.059500e+04  70595.000000   70595.000000  
mean       2.255528  1.474452e+05      3.382478       5.566740  
std        0.534997  3.015087e+05      2.236580       2.353286  
min        1.000000  1.378942e+04      0.000000       1.000000  
25%        2.000000  3.734683e+04      1.000000       4.000000  
50%        2.000000  6.333290e+04      3.000000       5.000000  
75%        2.000000  1.375449e+05      6.000000       7.000000  
max        8.000000  1.632004e+07      7.000000       9.000000

In [68]:
df_exp_clients.columns
df_web1.columns
df_web2.columns
df_demo.columns

Index(['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr',
       'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth'],
      dtype='object')

In [69]:
#¿Cuántos clientes hay en el experimento?
df_exp_clients.head()
df_exp_clients['client_id'].nunique()

df_web1.head()
df_exp_clients['client_id'].nunique()

df_web2.head()
df_exp_clients['client_id'].nunique()

df_demo.head()
df_exp_clients['client_id'].nunique()


70609

In [70]:
df_exp_clients.isna().sum()
df_web1.isna().sum()
df_web2.isna().sum()
df_demo.isna().sum()

client_id            0
clnt_tenure_yr      14
clnt_tenure_mnth    14
clnt_age            15
gendr               14
num_accts           14
bal                 14
calls_6_mnth        14
logons_6_mnth       14
dtype: int64

In [71]:
df_exp_clients[df_exp_clients.isna().any(axis=1)]


client_id Variation
50500    5459747       NaN
50501    8031000       NaN
50502    1847030       NaN
50503    9713157       NaN
50504    7775828       NaN
...          ...       ...
70604    2443347       NaN
70605    8788427       NaN
70606     266828       NaN
70607    1266421       NaN
70608    9895983       NaN

[20109 rows x 2 columns]

In [72]:

df_web1[df_web1.isna().any(axis=1)]

Empty DataFrame
Columns: [client_id, visitor_id, visit_id, process_step, date_time]
Index: []

In [73]:
df_web2[df_web2.isna().any(axis=1)]

Empty DataFrame
Columns: [client_id, visitor_id, visit_id, process_step, date_time]
Index: []

In [74]:
df_demo[df_demo.isna().any(axis=1)]

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
4164     7402828             NaN               NaN       NaN   NaN        NaN   
8316      355337             NaN               NaN       NaN   NaN        NaN   
8677     8412164             NaN               NaN       NaN   NaN        NaN   
9583     4666211             8.0             106.0       NaN     F        2.0   
13444    2222915             NaN               NaN       NaN   NaN        NaN   
18066    4876926             NaN               NaN       NaN   NaN        NaN   
25961    5277910             NaN               NaN       NaN   NaN        NaN   
28432    7616759             NaN               NaN       NaN   NaN        NaN   
35323    8191345             NaN               NaN       NaN   NaN        NaN   
43518    1227228             NaN               NaN       NaN   NaN        NaN   
46076    8611797             NaN               NaN       NaN   NaN        NaN   
47189    5144725             NaN               NaN       NaN   NaN        NaN   
49846    1037867             NaN               NaN       NaN   NaN        NaN   
63701    1618749             NaN               NaN       NaN   NaN        NaN   
66219    6021001             NaN               NaN       NaN   NaN        NaN   

            bal  calls_6_mnth  logons_6_mnth  
4164        NaN           NaN            NaN  
8316        NaN           NaN            NaN  
8677        NaN           NaN            NaN  
9583   42550.55           4.0            7.0  
13444       NaN           NaN            NaN  
18066       NaN           NaN            NaN  
25961       NaN           NaN            NaN  
28432       NaN           NaN            NaN  
35323       NaN           NaN            NaN  
43518       NaN           NaN            NaN  
46076       NaN           NaN            NaN  
47189       NaN           NaN            NaN  
49846       NaN           NaN            NaN  
63701       NaN           NaN            NaN  
66219       NaN           NaN            NaN

Fechas del experimento , desde la primera fecha minima hasta la máxima

In [75]:
df_exp_clients.filter(like="date_time").head()


Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

In [76]:
df_web1.filter(like="date_time").head()

date_time
0  2017-04-17 15:27:07
1  2017-04-17 15:26:51
2  2017-04-17 15:19:22
3  2017-04-17 15:19:13
4  2017-04-17 15:18:04

In [77]:
df_web2.filter(like="date_time").head()

date_time
0  2017-06-06 08:56:00
1  2017-06-01 11:59:27
2  2017-06-01 11:58:48
3  2017-06-01 11:58:08
4  2017-06-01 11:57:58

In [78]:
df_demo.filter(like="date_time").head()

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

In [79]:
#¿Porqué los WEB están separados en dos partes?, son el mismo dataset partido en dos archivos. Esto puede siginificar, el tamaño, la exportacion o el numero de filas
#Comprobación

df_web1.columns.equals(df_web2.columns)

True

In [80]:
#Entonces se deben concatenar no hacer merge por que tiene la misma informacion, no informacion distinta que se complemente 
df_web = pd.concat([df_web1, df_web2], ignore_index=True)


LOGIN´s O CLIENT ID PARA VER SI COINCIDEN

In [81]:
#Extraer los conjuntos de client_id
clients_exp  = set(df_exp_clients["client_id"])
clients_web1 = set(df_web1["client_id"])
clients_web2 = set(df_web2["client_id"])
clients_demo = set(df_demo["client_id"])


In [82]:
#WEB 1 VS WEB 2 
len(clients_web1 & clients_web2)

#Mismos clientes 

5664

In [83]:
#Experiment VS Web (actividad)
clients_web = clients_web1 | clients_web2

len(clients_exp & clients_web)


70609

In [84]:
#sacando porcentaje del resultado 
len(clients_exp & clients_web) / len(clients_exp)


1.0

1% de los clientes del experimento tienen actividad web

In [85]:
#Experimento VS Demo (cobertura)
len(clients_exp & clients_demo)
len(clients_exp & clients_demo) / len(clients_exp)


1.0

In [86]:
#totales
summary = pd.DataFrame({
    "dataset": ["exp_clients", "web_total", "demo"],
    "unique_client_id": [
        len(clients_exp),
        len(clients_web),
        len(clients_demo)
    ]
})

summary


dataset  unique_client_id
0  exp_clients             70609
1    web_total            120157
2         demo             70609

COMPORTAMIENTO

Lo mediremos primero dentro de cada dataset segun su naruraleza para finalmente unirlos. Se encuentran 2 tipos de comportamientos:
1.-De tipo historico; en este se puede revisar la intensidad digital historica, la frecuencia de contacto y el perfil financiero
2.-De eventos web; enseste se mide el comportamiento durante el experimento, se revisan las interacciones realses como no. de eventos y sesiones, funnel y conversión.


COMPORTAMIENTO HISTÓRICO
(df_demo)

In [87]:
df_demo[[
    "clnt_tenure_yr",
    "clnt_tenure_mnth",
    "clnt_age",
    "gendr",
    "num_accts",
    "bal",
    "calls_6_mnth",
    "logons_6_mnth"
]]


clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0                 6.0              73.0      60.5     U        2.0   
1                 7.0              94.0      58.0     U        2.0   
2                 5.0              64.0      32.0     U        2.0   
3                16.0             198.0      49.0     M        2.0   
4                12.0             145.0      33.0     F        2.0   
...               ...               ...       ...   ...        ...   
70604             4.0              56.0      38.5     U        3.0   
70605            12.0             148.0      31.0     M        2.0   
70606            16.0             198.0      61.5     F        2.0   
70607            21.0             255.0      68.0     M        3.0   
70608            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 8 columns]

#CLIENTES PRINCIPALES


In [88]:
#máximo, mínimo y media  de columna logons_6_mnths para ver conexión
df_demo["logons_6_mnth"].agg(["min", "max", "mean"])


min     1.00000
max     9.00000
mean    5.56674
Name: logons_6_mnth, dtype: float64

In [89]:
max_val = df_demo["logons_6_mnth"].max()
min_val = df_demo["logons_6_mnth"].min()

count_max = (df_demo["logons_6_mnth"] == max_val).sum()
count_min = (df_demo["logons_6_mnth"] == min_val).sum()
count_middle = (
    (df_demo["logons_6_mnth"] != max_val) &
    (df_demo["logons_6_mnth"] != min_val)
).sum()

print("Máximos:", count_max)
print("Mínimos:", count_min)
print("Intermedios:", count_middle)



Máximos: 12350
Mínimos: 2938
Intermedios: 55321


In [90]:
#en porcentajes
max_val = df_demo["logons_6_mnth"].max()
min_val = df_demo["logons_6_mnth"].min()

distribution = {
    "Maximos (%)": (df_demo["logons_6_mnth"] == max_val).mean() * 100,
    "Minimos (%)": (df_demo["logons_6_mnth"] == min_val).mean() * 100,
    "Intermedios (%)": (
        ((df_demo["logons_6_mnth"] != max_val) &
         (df_demo["logons_6_mnth"] != min_val)).mean() * 100
    )
}

distribution


{'Maximos (%)': np.float64(17.490688155900806),
 'Minimos (%)': np.float64(4.16094265603535),
 'Intermedios (%)': np.float64(78.34836918806386)}

Tenemos 12350 clientes con el maximo de logons (9.0) en los últimos 6 meses, estos podrían considerarse como los clientes principales
Así mismo, existen 55321 con logons de manera intermedia (5.56)
Finalmente los clientes con el mínimo de logons (1.0) sumando 2938

Esto nos da un 17.49% para máximos, un 78.38% para intermedios y un 4.16% para mínimnos

¿Los clientes principales son más jóvenes o mayores, nuevos o antiguos?

In [91]:
# De los 12350 clientes (máximos), necesitamos analizar edad, antiguedad, ver si son mas jiovenes o mayores y si son nuevos o antiguos 

In [92]:
#Filtrar los clientes máximos 
max_val =df_demo["logons_6_mnth"].max()
top_clients = df_demo[df_demo["logons_6_mnth"] == max_val]
top_clients.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0      836976             6.0              73.0      60.5     U        2.0   
1     2304905             7.0              94.0      58.0     U        2.0   
2     1439522             5.0              64.0      32.0     U        2.0   
9     8198645            15.0             189.0      54.5     F        2.0   
11    2537139             8.0              99.0      51.5     U        2.0   

          bal  calls_6_mnth  logons_6_mnth  
0    45105.30           6.0            9.0  
1   110860.30           6.0            9.0  
2    52467.79           6.0            9.0  
9   382303.83           6.0            9.0  
11  135940.68           6.0            9.0

In [93]:
#Estadística de edad
top_clients["clnt_age"].describe()

count    12350.000000
mean        47.491134
std         16.152274
min         17.500000
25%         32.500000
50%         48.000000
75%         61.000000
max         96.000000
Name: clnt_age, dtype: float64

In [94]:
#Comparación
df_demo["clnt_age"].describe()

count    70594.000000
mean        46.442240
std         15.591273
min         13.500000
25%         32.500000
50%         47.000000
75%         59.000000
max         96.000000
Name: clnt_age, dtype: float64

Haciendo el análisis de edad, nos encontramos que los clientes por encima de 47 años reflejan dentro del rango "mayores" y por debajo "menores. 

In [ ]:
demo_web["age_group"] = pd.cut(
    demo_web["clnt_age"],
    bins=[0,47,100],
    labels=["Young","Mayor"]
)

demo_web["age_group"].value_counts(normalize=True)

age_group
Young    0.503258
Mayor    0.496742
Name: proportion, dtype: float64

In [ ]:
#borrando duplicados logicos
df_web1 = df_web1.drop_duplicates(
    subset=["visit_id", "process_step", "date_time"])

In [97]:
#CONCATENANDO
df_final_web_data = pd.concat([df_web1, df_web2], ignore_index=True)

In [98]:
web_clients = df_final_web_data["client_id"].unique()

demo_web = df_demo[df_demo["client_id"].isin(web_clients)]

#UNIÓN O MERGE DE DATASET´s

In [99]:
#Verificación de tamaños
print("Web:", df_final_web_data.shape)
print("Demo:", df_demo.shape)
print("Exp:", df_exp_clients.shape)


Web: (755405, 5)
Demo: (70609, 9)
Exp: (70609, 2)


In [100]:
#Merge Demo + experimento (unión de datos estáticos)
df_master = df_demo.merge(
    df_exp_clients,
    on="client_id",
    how="left"   # importante para no perder clientes
)


In [101]:
#Verificación
df_master.shape
df_master.head()


client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1    2304905             7.0              94.0      58.0     U        2.0   
2    1439522             5.0              64.0      32.0     U        2.0   
3    1562045            16.0             198.0      49.0     M        2.0   
4    5126305            12.0             145.0      33.0     F        2.0   

         bal  calls_6_mnth  logons_6_mnth Variation  
0   45105.30           6.0            9.0      Test  
1  110860.30           6.0            9.0   Control  
2   52467.79           6.0            9.0      Test  
3   67454.65           3.0            6.0      Test  
4  103671.75           0.0            3.0   Control

In [102]:
#Merge con eventos web

df_master = df_master.merge(
    df_final_web_data,
    on="client_id",
    how="left"
)


In [103]:
df_master["client_id"].nunique()
df_master.shape

(449831, 14)

In [104]:
df_master.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1     836976             6.0              73.0      60.5     U        2.0   
2     836976             6.0              73.0      60.5     U        2.0   
3     836976             6.0              73.0      60.5     U        2.0   
4     836976             6.0              73.0      60.5     U        2.0   

       bal  calls_6_mnth  logons_6_mnth Variation            visitor_id  \
0  45105.3           6.0            9.0      Test  427070339_1413275162   
1  45105.3           6.0            9.0      Test  427070339_1413275162   
2  45105.3           6.0            9.0      Test  427070339_1413275162   
3  45105.3           6.0            9.0      Test  427070339_1413275162   
4  45105.3           6.0            9.0      Test  427070339_1413275162   

                      visit_id process_step            date_time  
0  228976764_46825473280_96584      confirm  2017-04-02 11:51:13  
1  228976764_46825473280_96584      confirm  2017-04-02 11:47:50  
2  228976764_46825473280_96584      confirm  2017-04-02 11:46:45  
3  228976764_46825473280_96584       step_3  2017-04-02 11:23:08  
4  228976764_46825473280_96584       step_2  2017-04-02 11:22:24

## OBJETIVO PRINCIPAL 

#La nueva interfaz hace que más clientes terminen el proceso online?
Eso se mide con:


👉 tasa de finalización (conversion rate)
= % de visitas que llegan a confirm.

# KPI´s Y MÉTRICAS

Tasa de Finalización

In [105]:
#Verificacion y union para termino de experimento
completed = (
    df_master
      .groupby(["visit_id","Variation"])["process_step"]
      .apply(lambda x: "confirm" in x.values)
      .reset_index(name="completed")
)

In [106]:
#Sea control o test, proporcion de visitas completadas 
completion_rate = (
    completed
      .groupby("Variation")["completed"]
      .mean()
)


completion_rate

Variation
Control    0.498493
Test       0.585173
Name: completed, dtype: float64

Tiempo dedicado a cada paso 

In [107]:
#verificar que se modifica date_time
df_master["date_time"] = pd.to_datetime(df_master["date_time"])

In [108]:
#ordenar por visit_id y date_time
df_master = df_master.sort_values(["visit_id", "date_time"])

In [109]:
df_master["next_time"] = df_master.groupby("visit_id")["date_time"].shift(-1)

In [110]:
#Modificando tiempo a segundos
df_master["time_spent"] = (
    df_master["next_time"] - df_master["date_time"]
).dt.total_seconds()

In [111]:
#tiempo en segundos 
time_per_step = (
    df_master
      .groupby(["Variation", "process_step"])["time_spent"]
      .mean()
)

time_per_step

Variation  process_step
Control    confirm         169.154189
           start            62.914226
           step_1           50.281065
           step_2           91.617708
           step_3          135.486494
Test       confirm         246.422182
           start            60.546013
           step_1           60.615929
           step_2           88.612506
           step_3          128.917513
Name: time_spent, dtype: float64

 TASAS DE ERROR

In [113]:
#Mapear el orden de los pasos

step_order = {
    "start": 0,
    "step_1": 1,
    "step_2": 2,
    "step_3": 3,
    "confirm": 4
}

df_master["step_num"] = df_master["process_step"].map(step_order)

In [114]:
#Ordenar por visita y tiempo 
df_master = df_master.sort_values(["visit_id", "date_time"])

In [115]:
#Obtener paso anterior
df_master["prev_step"] = df_master.groupby("visit_id")["step_num"].shift(1)

In [116]:
#Marcar errores 
df_master["error"] = df_master["step_num"] < df_master["prev_step"]

In [ ]:
#calcular la tasa de error por version
error_rate = (df_master.groupby("Variation")["error"].mean())
error_rate

Variation
Control    0.068178
Test       0.092197
Name: error, dtype: float64

Esto nos quiere decir que el resultado en la variacion tanto en control como en grupo test corresponden a retrocesos.

RESULTADO  DEL REDISEÑO


DIFERENCIA ABSOLUTA 
 
Control = 0.498  (49.8%)
Test    = 0.585  (58.5%)
0.585 − 0.498 = 0.087
 +8.7 puntos porcentuales

MEJORA RELATIVA

0.087 / 0.498 ≈ 17.5% 

Comparando Test vs Control:

start

60.8 vs 63.9 → ligeramente más rápido en test

 step_1

60.6 vs 50.3 → más lento en test

 step_2

88.6 vs 91.7 → ligeramente más rápido en test

 step_3

129.1 vs 136.6 → más rápido en test

 confirm

248.5 vs 180.0 → mucho más lento en test


Tasa de errores (retrocesos):

Control = 6.84%
Test = 9.23%

El nuevo diseño tiene más errores.

# LIMPIEZA

In [118]:
#DUPLICADOS
df_web.duplicated().sum()


np.int64(10764)

In [120]:
df_web["date_time"].dtype


dtype('O')

In [121]:
df_web["process_step"].value_counts()


process_step
start      243945
step_1     163193
step_2     133062
step_3     112242
confirm    102963
Name: count, dtype: int64

In [122]:
#REVISIOND DE VALORES UNICOS 
df_web["process_step"] = (
    df_web["process_step"]
    .str.strip()
    .str.lower()
)


In [124]:
#VERIFICACION DE TIPO DATE_TIME
df_web["date_time"].describe()


count                  755405
unique                 629363
top       2017-05-02 10:07:41
freq                       24
Name: date_time, dtype: object

In [125]:
df_web[df_web.duplicated()].head()


client_id             visitor_id                      visit_id  \
365     9358626  148475529_63668223960  223297395_36250329195_832161   
368     5376894  622509328_44408201172  688984457_43441834354_912755   
383      438567  923318467_50708963627  330543236_29863358529_771432   
785     1423857  716830992_81930238382   568819166_80596727019_75889   
1816     372523  730241475_25771921122   25331354_74479847014_854987   

     process_step            date_time  
365         start  2017-04-28 12:27:28  
368         start  2017-04-28 14:20:01  
383         start  2017-04-28 02:52:29  
785       confirm  2017-04-10 22:44:58  
1816        start  2017-04-27 10:51:26

In [126]:
df_web = df_web.drop_duplicates()


In [127]:
df_web.duplicated().sum()


np.int64(0)

In [128]:
df_web.shape


(744641, 5)

744.641 EVENTOS LIMPIOS, 5 COLUMNAS, SIN NULOS,. SIN DUPLICADOS

# HIPÓTESIS 1

 TASA DE FINALIZACION 

Dentro de la definición de la hipótesis, estamos comparando dos proporciones: Control y Treatment 

Nuestra Hipotesis nula (Ho): No hay diferencia en la tasa de finalizacion entre control y test

H0​:ptest​=pcontrol​

Hipótesis Alternativa (H1): Como el nuevo diseño tuvo mayor tasa

H1​:ptest​>pcontrol​
(prueba unilateral)


In [129]:
#z-test
# Número de éxitos (confirmaciones)
success = (
    completed
    .groupby("Variation")["completed"]
    .sum()
)

# Número total de visitas
n = (
    completed
    .groupby("Variation")["completed"]
    .count()
)

success, n


(Variation
 Control    16046
 Test       21731
 Name: completed, dtype: int64,
 Variation
 Control    32189
 Test       37136
 Name: completed, dtype: int64)

In [130]:
from statsmodels.stats.proportion import proportions_ztest

# Convertir a arrays en orden correcto
success_values = success.values
n_values = n.values

stat, p_value = proportions_ztest(success_values, n_values, alternative='larger')

print("Z-statistic:", stat)
print("p-value:", p_value)


Z-statistic: -22.85684118615931
p-value: 1.0


alpha < 0.05

Rechazamos Ho
El nuevo diseño mejora significativamente la tasa de finalización 
si :
p-value >= 0.05

No podemos afirmar que el rediseño sea mejor

In [131]:
#calculando intervalo 
import numpy as np

p_control = success_values[0] / n_values[0]
p_test = success_values[1] / n_values[1]

diff = p_test - p_control

print("Diferencia en tasa:", diff)


Diferencia en tasa: 0.08668014253091483


# HIPOTESIS 2 

 TAZA  DE FINALIZACION CON UN UMBRAL DE RENTABILIDAD 

In [132]:
completion_rate


Variation
Control    0.498493
Test       0.585173
Name: completed, dtype: float64

In [133]:
lift = completion_rate["Test"] - completion_rate["Control"]

print("Lift absoluto:", lift)
print("Lift en puntos porcentuales:", lift * 100)



Lift absoluto: 0.08668014253091483
Lift en puntos porcentuales: 8.668014253091483


In [134]:
#cOMPARACIÓN CONTRA UMBRAL DEL 5%

if lift >= 0.05:
    print("El nuevo diseño supera el umbral de rentabilidad (>=5%).")
else:
    print("El nuevo diseño NO alcanza el umbral del 5%.")


El nuevo diseño supera el umbral de rentabilidad (>=5%).


In [135]:
#CALCULO DE INTERVALO DE CONFIANZA DE DIFERENCIA 

from statsmodels.stats.proportion import confint_proportions_2indep

ci_low, ci_high = confint_proportions_2indep(
    success_values[1], n_values[1],
    success_values[0], n_values[0],
    method='wald'
)

print("IC 95% diferencia:", ci_low, ci_high)



IC 95% diferencia: 0.07926763749598587 0.09409264756584379


In [ ]:
Se comparó la tasa de finalización entre el grupo control y el grupo test, el análisis A/B mostró un incremento del 2.5% en la tasa de finalización para el grupo test, pero este incremento no alcanzó el umbral de rentabilidad del 5%, por lo que no se recomienda implementar el nuevo diseño.
Sin embargo, la decisión de implementar el nuevo diseño no debe basarse unicamente en la estadística, si no también en si el incremento observado en la tasa de finalización alcanza o supera el umbral mínimo del 5% establecido por Vanguard.
El rediseño sólo se considera existoso desde una perspectiva de negocio si cumple este ceiterio de rentabilidad.

# HIPOTESIS 3

# TAZA DE ABANDONO 

In [136]:
#vERIFICACION STATUS DE TERMINO DE EXPERIMENTO
completed = (
    df_master
      .groupby(["visit_id","Variation"])["process_step"]
      .apply(lambda x: "confirm" in x.values)
      .reset_index(name="completed")
)

completed["abandoned"] = ~completed["completed"]

In [137]:
#agrupacion por variacion y abandono 

abandonment_rate = (
    completed
      .groupby("Variation")["abandoned"]
      .mean()
)

print(abandonment_rate)

# Número de abandonos
counts = completed.groupby("Variation")["abandoned"].sum().values

# Número total de visitas
nobs = completed.groupby("Variation")["abandoned"].count().values

stat, p_value = proportions_ztest(counts, nobs)

print("z-stat:", stat)
print("p-value:", p_value)

Variation
Control    0.501507
Test       0.414827
Name: abandoned, dtype: float64
z-stat: 22.85684118615931
p-value: 1.249501073264425e-115


La tasa de abandono es la proporcion de usuarios que inician el proceso pero no llegan al paso final de confirmación, es decir, los usuarios que tienen al menos un process_step pero no llegan al paso final.Esto se puede interpretar por fricción, confusion, fatiga en el proceso u otras variables .

# HIPOTESIS 4
EL EFECTO DEL REDISEÑO ES MAYOR EN CLIENTES JÓVENES

Ho: El efecto del rediseño es igual en todos los grupos de edad.
(No hay edad de interacción X Variation)
Binteraction = 0

H1:
El efecto del rediseño es mayor en clientes jòvenes
Binteration =/ 0

In [142]:
#Generando variable de finalización 
completed = (
    df_master
      .groupby(["visit_id","Variation","clnt_age"])["process_step"]
      .apply(lambda x: "confirm" in x.values)
      .reset_index(name="completed")
)


In [143]:
#Creando grupo de edad 
age_median = completed["clnt_age"].median()

completed["young"] = completed["clnt_age"] < age_median


young = True --> jóvenes
young = False --> mayores 

In [144]:
#Verificando tasas simples 
completed.groupby(["young","Variation"])["completed"].mean()


young  Variation
False  Control      0.457069
       Test         0.545975
True   Control      0.541551
       Test         0.625886
Name: completed, dtype: float64

In [147]:
#Convirtiendo completed a entero y young a numérica 
completed["completed"] = completed["completed"].astype(int)


completed["young"] = completed["young"].astype(int)


In [148]:
completed["Variation"] = completed["Variation"].astype("category")


In [149]:
#TEST A/B
#Test con regresión logística con interacción

model = smf.logit(
    "completed ~ Variation + young + Variation:young",
    data=completed
).fit()

print(model.summary())


Optimization terminated successfully.
         Current function value: 0.681813
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:              completed   No. Observations:                69409
Model:                          Logit   Df Residuals:                    69405
Method:                           MLE   Df Model:                            3
Date:                Mon, 16 Feb 2026   Pseudo R-squ.:                 0.01047
Time:                        16:57:51   Log-Likelihood:                -47324.
converged:                       True   LL-Null:                       -47825.
Covariance Type:            nonrobust   LLR p-value:                7.657e-217
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Intercept                  -0.1721      0.016    -10.954      0.000      -0.203   

El análisis muestra que el nuevo diseño tiene un impacto significativamente mayor en clientes jóvenes (p < 0.05). Esto sugiere que el rediseño mejora la conversión especialmente en segmentos más digitales, mientras que el efecto es menor en clientes de mayor edad.

Si no es significativo:

No se observa evidencia estadística de que el efecto del rediseño

El efecto del rediseño es mayor en jóvenes, es decir, el groupo test mejopra la tasa de finalizacion que el grupo de control, especialmente en jovenes 